docs/architecture/adr-022-multi-versioned-state-machine.md

```diff
@@ -0,0 +1,98 @@
+
+## Detailed Design
+
+There are several parts of the protocol that were modified.
+
+### Loading and Setting the App Version
+
+The app version was previously never persisted to disk, but was simply hardcoded into the application (as 1). An application with multiple versions needs to be able to persist them to disk and load them in the case the node crashes or is shutdown. The `Info` abci call must pass the app version to Comet for Comet to generate headers with it. The decision was made to always require the app version to be set in the consensus params of Genesis (prior `ConsensusParams` wasn't mandatory). Upon first use, `Info` will return 0 (as no app version has been persisted) and then take the app version passed in `InitChain` to initialize the app version. It is persisted both to disk in the `ParamStore` and in memory in the `BaseApp` struct. Upon later startups, `Info` will load the app version from disk, set it in memory and also return it to Comet. The `AppVersion` is only later adjusted when `Upgrade` is called on the `Application`. Upgrading is no longer a module but is more a native part of the application. Modules register migrations and `Upgrade`, which is executed in `EndBlock` will trigger those migrations and then set the new version. Upgrading needs to happen in `EndBlock`, such that `PrepareProposal` for the following height already is run with the logic of the new version.
+
+### SDK Context
+
+The `sdk.Context` type is passed to almost every component in the application and is thus ideal for storing and accessing the current app version. It already contains the app version in the block header. However this is not always populated. Thus `InitChain` and `PrepareProposal` were modified to include the app version as either passed from the genesis file or loaded from the application itself.
+
+### Module Manager Changes
+
+The module manager was modified to contain a mapping of modules to their version.
+
+```go
+type Manager struct {
+    versionedModules map[uint64]map[string]sdkmodule.AppModule
+    ...
+}
+```
+
+Each module has a `Name` and a `ConsensusVersion` that marks it unique. No two modules can have the same `Name` within a single app version. When constructing the application, we specify the contiguous range (inclusive) of versions that the module is part of:
+
+```go
+app.mm, err = module.NewManager([]module.VersionedModule{
+    {
+        Module:      blobstream.NewAppModule(appCodec, app.BlobstreamKeeper),
+        FromVersion: v1, ToVersion: v2,
+    },
+    {
+        Module:      upgrade.NewAppModule(app.UpgradeKeeper),
+        FromVersion: v2, ToVersion: v2,
+    },
+})
+```
+
+`InitGenesis`, `BeginBlock` and `EndBlock` will only be called for the modules that belong to the current version of the application.
```

Comment:
Is `InitGenesis` invoked for subsequent app versions (i.e. after app version 1)?

Reply:
InitGenesis at the module manager level is called once for all modules of that version. However when we upgrade, if we have a new module (for example minfee) we call InitGenesis on specifically the module with the DefaultGenesis of the module